### PR TITLE
 Various code quality improvements and added TODOs for suggestions for security and code quality

### PR DESCRIPTION
**Pull request by jamesdwilson**
Various code quality improvements and added TODOs for suggestions for security and code quality

### CODE REVIEW

File: js/views/itemEditVw.js [hunk 1]
```diff
@@ -111,7 +111,7 @@ module.exports = Backbone.View.extend({
     if(priceType === "local"){
       inputParent.find('.js-priceLocal').val(inputParent.find('.js-priceBtc').val() * window.currentBitcoin);
     }else{
-      inputParent.find('.js-priceBtc').val(inputParent.find('.js-priceLocal').val() / window.currentBitcoin);
+      inputParent.find('.js-priceBtc').val(inputParent.find('.js-priceLocal').val() / window.currentBitcoin);  //TOOD: Eliminate division because of floating point math
```
Comment (jjeffryes) [inline js/views/itemEditVw.js:114]: The floating point math isn't an issue here, since the result is a rounded up approximation.

File: js/views/itemEditVw.js [hunk 2]
```diff
@@ -143,7 +143,7 @@ module.exports = Backbone.View.extend({
 
   changeType: function(e) {
     var typeValue = $(e.target).val();
-    if(typeValue === "physical good") {
+    if(typeValue === "physical good") { //TODO: remove magic text, make constant or lookup
```
Comment (jjeffryes) [inline js/views/itemEditVw.js:146]: In this case, the value is what's set by the API, we don't need to over-optimize here.

File: js/views/itemEditVw.js [hunk 3]
```diff
@@ -160,23 +160,26 @@ module.exports = Backbone.View.extend({
...
       data: formData,
-      success: function(data) {
+      success: function(data) { //TODO: Have JQuery parse the JSON directly in ajax call
+        var errorModal, hashArray, imageArray;
         data = JSON.parse(data);
+
         if (data.success === true){
-          var imageArray = __.clone(self.model.get("combinedImagesArray"));
-          var hashArray = __.clone(self.model.get("imageHashesToUpload"));
+          imageArray = __.clone(self.model.get("combinedImagesArray"));
```
Comment (jjeffryes) [inline js/views/itemEditVw.js:170]: Why did you remove var here? This will hoist imageArray to the global scope, or any parent scope with an imageArray variable. we want to keep it in the success function.

File: js/views/itemEditVw.js [hunk 4]
```diff
@@ -160,23 +160,26 @@ module.exports = Backbone.View.extend({
...
+          hashArray = __.clone(self.model.get("imageHashesToUpload"));
```
Comment (jjeffryes) [inline js/views/itemEditVw.js]: See above.

File: js/views/itemEditVw.js [hunk 5]
```diff
@@ -160,23 +160,26 @@ module.exports = Backbone.View.extend({
...
           __.each(data.image_hashes, function (hash) {
-            imageArray.push(self.model.get('server') + "get_image?hash=" + hash);
+            imageArray.push(self.model.get('server') + "get_image?hash=" + hash); //TODO: Change to more formal URL reference from config file
```
Comment (jjeffryes) [inline js/views/itemEditVw.js:173]: This TODO isn't needed, the model serves as the config file, it stores the local settings.

File: js/views/itemEditVw.js [hunk 6]
```diff
@@ -160,23 +160,26 @@ module.exports = Backbone.View.extend({
...
             hashArray.push(hash);
-          })
+          });
           self.model.set("combinedImagesArray", imageArray);
           self.model.set("imageHashesToUpload", hashArray);
 
           self.updateImages();
         }else if (data.success === false){
-          var errorModal = $('.js-messageModal');
+          errorModal = $('.js-messageModal');
```
Comment (jjeffryes) [inline js/views/itemEditVw.js:181]: Don't remove var. This is a local variable.

File: js/views/itemEditVw.js [hunk 7]
```diff
@@ -217,19 +220,33 @@ module.exports = Backbone.View.extend({
   },
 
   deleteImage: function(e) {
-    var imgIndex = $(e.target).closest('.itemImg').data('index');
+    var imageUploadArray, imgIndex = $(e.target).closest('.itemImg').data('index');
```
Comment (jjeffryes) [inline js/views/itemEditVw.js:225]: If you're going to declare variables together, put a return after the , so it's clear they're separate. 

In general, declaring variables near where they are used is deliberate, to make the code more readable for future contributors. While that could create a hoisting issue, we consider that an ok tradeoff.

File: js/views/itemEditVw.js (hunk 3 above)
Comment (jamesdwilson) [inline js/views/itemEditVw.js:170]: I moved them up to line 164.

Comment (jjeffryes) [inline js/views/itemEditVw.js:170]: Sorry, ignore the above, I missed the declaration at the top.

File: js/views/itemEditVw.js [hunk 8]
```diff
@@ -160,23 +160,26 @@ module.exports = Backbone.View.extend({
       contentType: false,
       processData: false,
       data: formData,
-      success: function(data) {
+      success: function(data) { //TODO: Have JQuery parse the JSON directly in ajax call
+        var errorModal, hashArray, imageArray;
```
Comment (jjeffryes) [inline js/views/itemEditVw.js]: Format this as:

var errorModal, 
      hashArray, 
      imageArray;

File: js/views/itemEditVw.js (hunk 6 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:181]: Ignore this one too.

Comment (jamesdwilson) [inline js/views/itemEditVw.js:181]: There are no such things as local variables in javascript, only function block variables. All variables are hoisted to the top of the function they are in.  I am declaring errorModal on line 164. All variable declarations should be at the top of the function.

File: js/views/itemEditVw.js (hunk 7 above)
Comment (jamesdwilson) [inline js/views/itemEditVw.js:225]: It is a bad practice IMO because not only of hoisting, but as I've already seen a couple times in the codebase, it can allow for the unawareness of duplicating the declaration, as well as when it is var'd again it will clobber the previous value. 
With strict mode, this is not possible.

File: js/views/itemEditVw.js (hunk 6 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:181]: By local I meant local to the function. 

See the note on why variables were not always declared at the top.

Comment (jamesdwilson) [inline js/views/itemEditVw.js:181]: Strict mode is a good practice and it will require variable declaration at the top, for a number of reasons, some of which I've outlined.

File: js/views/userPageVw.js [hunk 9]
```diff
@@ -389,13 +391,10 @@ module.exports = Backbone.View.extend({
...
         }else if (data.success === false){
-          var errorModal = $('.js-messageModal');
-          errorModal.removeClass('hide');
+          //TODO: Extract these lines to a more generic showModal-type function
           errorModal.find('.js-messageModal-title').text("Changes Could Not Be Saved");
           errorModal.find('.js-messageModal-message').html("Uploading the image has failed due to the following error: <br/><br/><i>" + data.reason + "</i>");
-        }else if (imageHash == "b472a266d0bd89c13706a4132ccfb16f7c3b9fcb") {
-          var errorModal = $('.js-messageModal');
-          errorModal.removeClass('hide');
+        }else if (imageHash == "b472a266d0bd89c13706a4132ccfb16f7c3b9fcb") { //TODO: Remove magic number
```
Comment (jjeffryes) [inline js/views/userPageVw.js:397]: This number is the hex code for "", to remove it we'd need to alter the API.

Comment (jamesdwilson) [inline js/views/userPageVw.js:397]: What? You're suggesting leaving in a raw reference to b472a266d0bd89c13706a4132ccfb16f7c3b9fcb?

File: js/views/userPageVw.js [hunk 10]
```diff
@@ -381,6 +381,8 @@ module.exports = Backbone.View.extend({
       processData: false,
       data: formData,
       success: function(data) {
+        var errorModal = $('.js-messageModal');
+        errorModal.removeClass('hide');
```
Comment (jjeffryes) [inline js/views/userPageVw.js:385]: Moving the declaration here is fine, but the errorModal should only be unhidden when one of the error conditions is met.

Comment (jamesdwilson) [inline js/views/userPageVw.js:385]: Good point. I saw it in the if and the else though, so I may have been mistaken thinking it should just be done arbitrarily.

File: js/views/itemEditVw.js (hunk 7 above)
Comment (hoffmabc) [inline js/views/itemEditVw.js:225]: When did we decide to move to strict mode?

File: js/views/userPageVw.js (hunk 9 above)
Comment (hoffmabc) [inline js/views/userPageVw.js:397]: This is just a hash for an empty string. Until the API is updated this is the way it needs to remain. @cpacia can elaborate if necessary.

File: js/views/userPageVw.js [hunk 11]
```diff
@@ -505,7 +504,7 @@ module.exports = Backbone.View.extend({
 
     $.ajax({
       type: "DELETE",
-      url: self.item.get('server') + "contracts/?id="+ self.item.get('id'),
+      url: self.item.get('server') + "contracts/?id="+ self.item.get('id'), //TODO: Change from GET to POST for safe methods, http://restcookbook.com/HTTP%20Methods/idempotency/
```
Comment (jjeffryes) [inline js/views/userPageVw.js:507]: What does this refer to?

The gets in the line with the TODO are Backbone get calls to the item model.

File: js/views/itemEditVw.js (hunk 7 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:225]: We didn't, it's a suggestion from James.

Comment (hoffmabc) [inline js/views/itemEditVw.js:225]: Just want to make sure this is an agreed upon move because it changes the approach and coding style.

Comment (jamesdwilson) [inline js/views/itemEditVw.js:225]: I think it is a smart move. Forgive me if I am out of line. I think it is industry standard and a healthy practice. It prevents a lot of common JS errors. Maybe you'll appreciate John Resig's views on it. http://ejohn.org/blog/ecmascript-5-strict-mode-json-and-more/ Or Douglas Crockford views on it as it was directly inspired by his book Javascript: The Good Parts, as well as from the YUI team he headed: http://yuiblog.com/blog/2010/12/14/strict-mode-is-coming-to-town/

File: js/views/userPageVw.js (hunk 11 above)
Comment (jamesdwilson) [inline js/views/userPageVw.js:507]: Please read the provided link to http://restcookbook.com/HTTP%20Methods/idempotency/ - it explains why you should not send destructive methods by GET.

File: js/views/itemEditVw.js (hunk 7 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:225]: It's a good practice, but we didn't start with it so we'll have to consider how much of an impact going back to refactor things would have before adding it in across the project. I think it would be fine to add as things are worked on.

File: js/views/userPageVw.js (hunk 11 above)
Comment (hoffmabc) [inline js/views/userPageVw.js:507]: This is a DELETE method what am I missing?

Comment (jamesdwilson) [inline js/views/userPageVw.js:507]: It should be called with the arguments in the post body

File: js/views/itemEditVw.js (hunk 8 above)
Comment (jamesdwilson) [inline js/views/itemEditVw.js]: Should I commit this change now or do you only mean this a suggestion for the future?

File: js/views/userPageVw.js (hunk 11 above)
Comment (jamesdwilson) [inline js/views/userPageVw.js:507]: See the table on the link provided, and see how DELETE is not a safe method, and should have the arguments in the POST body.

Comment (jjeffryes) [inline js/views/userPageVw.js:507]: DELETE, as used here, is fine. If you want to modify the API to accept a POST with a parameter that would delete the contract instead, you can put in a pull request that does that.

Comment (jamesdwilson) [inline js/views/userPageVw.js:507]: Why is it fine?

File: js/views/itemEditVw.js (hunk 8 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js]: Go ahead and do this now.

File: js/views/itemEditVw.js (hunk 2 above)
Comment (jamesdwilson) [inline js/views/itemEditVw.js:146]: What do you think about having the values shared in a common config file?

File: js/views/itemEditVw.js (hunk 8 above)
Comment (jamesdwilson) [inline js/views/itemEditVw.js]: pushed a5ee86fac60a97fed53f7782683891b29d2ced36

File: js/views/itemEditVw.js (hunk 2 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:146]: The only time this value would change is if the API also changed. I don't think putting it in a config file would provide a benefit. 

In general, the user model acts as a config file, and settings like the server URL and the SSL flag are (or will be) stored there.

File: js/views/userPageVw.js (hunk 11 above)
Comment (jjeffryes) [inline js/views/userPageVw.js:507]: As stated, this is how the API works. It expects a DELETE.

File: js/views/itemEditVw.js (hunk 1 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:114]: Remove this TODO.

File: js/views/itemEditVw.js (hunk 2 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:146]: Remove this TODO.

File: js/views/itemEditVw.js (hunk 5 above)
Comment (jjeffryes) [inline js/views/itemEditVw.js:173]: Remove this TODO.

File: js/views/userPageVw.js (hunk 9 above)
Comment (jjeffryes) [inline js/views/userPageVw.js:397]: Remove this TODO.

File: js/views/userPageVw.js (hunk 10 above)
Comment (jjeffryes) [inline js/views/userPageVw.js:385]: Move the removeClass back to inside the if statements.

File: js/views/userPageVw.js (hunk 11 above)
Comment (jjeffryes) [inline js/views/userPageVw.js:507]: Remove this TODO.

**Comment by jjeffryes** [comment]
Thank you for working on this, James. Please make the corrections requested, and re-submit the pull request.
